File: Lecture 28/Python Code/BuilderPattern.py
# check pythonic version of immutablility
# dataclass and frozen=True
# property decorator for getters

# Target Class
class HttpRequest:
    def __init__(self):
        self.url = None
        self.method = None
        self.headers = {}
        self.query_params = {}
        self.body = ""
        self.timeout = None

    def execute(self):
        if not self.url:
            raise ValueError("URL cannot be empty")

        print(f"Executing {self.method} request to {self.url}")

        if self.query_params:
            print("Query Parameters:")
            for k, v in self.query_params.items():
                print(f"  {k}={v}")

        if self.headers:
            print("Headers:")
            for k, v in self.headers.items():
                print(f"  {k}: {v}")

        if self.body:
            print(f"Body: {self.body}")

        print(f"Timeout: {self.timeout} seconds")
        print("Request executed successfully!")

# Builder Class
class HttpRequestBuilder:
    def __init__(self):
        self.req = HttpRequest()

    # Method chaining with reference return type so that we can chain multiple calls
    def with_url(self, url):
        self.req.url = url
        return self

    def with_method(self, method):
        self.req.method = method
        return self

    def with_header(self, key, value):
        self.req.headers[key] = value
        return self

    def with_query_param(self, key, value):
        self.req.query_params[key] = value
        return self

    def with_body(self, body):
        self.req.body = body
        return self

    def with_timeout(self, timeout):
        self.req.timeout = timeout
        return self

    # Build method to create the immutable HttpRequest object
    # Performs validation at one place --> Resolves the Scatter-Validation problem
    def build(self):
        # Validation logic can be added here
        if not self.req.url:
            raise ValueError("URL cannot be empty")
        if not self.req.method:
            self.req.method = "GET"  # Default method
        if not self.req.timeout:
            self.req.timeout = 30  # Default timeout
        return self.req


# --- Usage ---
if __name__ == "__main__":

    # Reason for using so many with_x methods with dot(.) --> because each method returns the builder object itself
    request = (
        HttpRequestBuilder()
        .with_url("https://api.example.com")
        .with_method("POST")
        .with_header("Content-Type", "application/json")
        .with_header("Accept", "application/json")
        .with_query_param("key", "12345")
        .with_body('{"name": "Aditya"}')
        .with_timeout(60)
        .build()    # Build method converts the builder object to the HttpRequest object at the end, 
                    # If we don't call build(), we will have a builder object instead of HttpRequest object
                    # Also called as Termination Method
    )

    request.execute()

    request2 = (
        HttpRequestBuilder()
        .with_url("https://api.example.com/users")
        .with_method("GET")
        .with_query_param("page", "1")
        .build()
    )

File: Lecture 28/C++ Code/BuilderPattern.cpp
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>
#include <map>

using namespace std;

class HttpRequest {
private:
    string url;
    string method;
    map<string, string> headers;
    map<string,string> queryParams;
    string body;
    int timeout; // in seconds

    // Private constructor - can only be accessed by the Builder
    HttpRequest() { }

public:
    friend class HttpRequestBuilder;  // Using friend class to allow access to private members

    // Method to execute the HTTP request
    void execute() {
        cout << "Executing " << method << " request to " << url << endl;
        
        if (!queryParams.empty()) {
            cout << "Query Parameters:" << endl;
            for (const auto& param : queryParams) {
                cout << "  " << param.first << "=" << param.second << endl;
            }
        }

        cout << "Headers:" << endl;
        for (const auto& header : headers) {
            cout << "  " << header.first << ": " << header.second << endl;
        }
        
        if (!body.empty()) {
            cout << "Body: " << body << endl;
        }
        
        cout << "Timeout: " << timeout << " seconds" << endl;
        cout << "Request executed successfully!" << endl;
    }
};

class HttpRequestBuilder {
private:    
    HttpRequest req;

public:    
    // Method chaining with reference return type so that we can chain multiple calls
    HttpRequestBuilder& withUrl(const string& u) {
        req.url = u; 
        return *this;
    }

    HttpRequestBuilder& withMethod(string method) {
        req.method = method;
        return *this;
    }
    
    HttpRequestBuilder& withHeader(const string& key, const string& value) {
        req.headers[key] = value;
        return *this;
    }

    HttpRequestBuilder& withQueryParams(const string& key, const string& value) {
        req.queryParams[key] = value;
        return *this;
    }
    
    HttpRequestBuilder& withBody(const string& body) {
        req.body = body;
        return *this;
    }
    
    HttpRequestBuilder& withTimeout(int timeout) {
        req.timeout = timeout;
        return *this;
    }
    
    // Build method to create the immutable HttpRequest object
    // Performs validation at one place --> Resolves the Scatter-Validation problem
    HttpRequest build() {
        // Validation logic can be added here
        if (req.url.empty()) {
            throw runtime_error("URL cannot be empty");
        }
        return req;
    }
};


int main() {
    // Using Builder Pattern (nested class)
    // Reason for using so many with_x methods with dot(.) --> because each method returns the builder object itself
    HttpRequest request = HttpRequestBuilder()
        .withUrl("https://api.example.com")
        .withMethod("POST")
        .withHeader("Content-Type", "application/json")
        .withHeader("Accept", "application/json")
        .withQueryParams("key", "12345")
        .withBody("{\"name\": \"Aditya\"}")
        .withTimeout(60)
        .build();   // Build method converts the builder object to the HttpRequest object at the end
                    // If we don't call build(), we will have a builder object instead of HttpRequest object
                    // Also called as Termination Method

    request.execute(); // Guaranteed to be in a consistent state


    // Another way of creating HttpRequest object
    HttpRequestBuilder* builder = new HttpRequestBuilder();
    HttpRequestBuilder builder2 = builder->withUrl("https://api.example.com/users");
    HttpRequestBuilder builder3 = builder2.withMethod("GET");
    HttpRequest request2 = builder3.build();    // Build method converts the builder object to the HttpRequest object at the end
}

